### PR TITLE
[WIP] FSEvent FileSystemWatcher port from CoreFX

### PR DESCRIPTION
**Pull request by alexischr**
* Depends on https://github.com/mono/mono/pull/4491 or your own build of System.Native dylib. While the above is being merged the easiest way to test the FSW is to first checkout the above PR, build, then switch to this one and build.
* The 'SR' class fields had to be manually extracted from CoreFX resx definitions, as otherwise they clashed with existing definitions from `ReferenceSources/SR.cs`
Behavior changes versus kqueue():
* no limit on watched files/directories, lower memory usage 
* renaming the watched directory will stop events from firing until renamed back; scope is path-string based. 
* Rapid-fire renames (A->B->C->D) may not all get an event but are 'eventually consistent' (you will see A and D)
* Mono's FSW (at least the kevent one) could watch a single file. This one only watches directories, and that also seems to be the .NET behavior.

### CODE REVIEW

**Comment by ludovic-henry** [comment]
Why did you copy the file over to mcs/class ? If there are some parts of the code from corefx that we do not want, we can put them between `#if !MONO ... #endif`

File: mcs/class/Interop/Interop.CoreFoundation.cs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        /// <param name="values">The values to put in the array</param>
+        /// <param name="numValues">The number of values in the array</param>
+        /// <param name="callbacks">Should be IntPtr.Zero</param>
+        /// <returns>Returns a pointer to a CFArray on success; otherwise, returns IntPtr.Zero</returns>
+        [DllImport(Interop.Libraries.CoreFoundationLibrary)]
+        private static extern SafeCreateHandle CFArrayCreate(
+            IntPtr allocator,
+            [MarshalAs(UnmanagedType.LPArray)]
+            IntPtr[] values,
+            ulong numValues,
```
Comment (alexischr) [inline mcs/class/Interop/Interop.CoreFoundation.cs]: Thanks!

Comment (Therzok) [inline mcs/class/Interop/Interop.CoreFoundation.cs]: https://github.com/dotnet/corefx/pull/15546

Submitted upstream

Comment (Therzok) [inline mcs/class/Interop/Interop.CoreFoundation.cs]: Newest corefx has the fix.

**Comment by alexischr** [comment]
Replaced by https://github.com/mono/mono/pull/5973

**Pull request closed without merging**